apps/server/src/modules/learnroom/service/board-copy.service.ts

```diff
@@ -23,53 +27,107 @@ export type BoardCopyParams = {
 @Injectable()
 export class BoardCopyService {
 	constructor(
+		private readonly logger: Logger,
+		private readonly boardRepo: BoardRepo,
 		private readonly taskCopyService: TaskCopyService,
 		private readonly lessonCopyService: LessonCopyService,
 		private readonly copyHelperService: CopyHelperService
 	) {}
 
 	async copyBoard(params: BoardCopyParams): Promise<CopyStatus> {
-		const elements: CopyStatus[] = [];
-		const references: BoardElement[] = [];
-
-		const boardElements = params.originalBoard.getElements();
-		for (let i = 0; i < boardElements.length; i += 1) {
-			const element = boardElements[i];
+		const { originalBoard, user, destinationCourse } = params;
 
-			if (element.boardElementType === BoardElementType.Task) {
-				const originalTask = element.target as Task;
-				const status = this.taskCopyService.copyTaskMetadata({
-					originalTask,
-					user: params.user,
-					destinationCourse: params.destinationCourse,
-				});
-				elements.push(status);
-				const taskBoardElement = BoardElement.FromTask(status.copyEntity as Task);
-				references.push(taskBoardElement);
-			} else if (element.boardElementType === BoardElementType.Lesson) {
-				const originalLesson = element.target as Lesson;
-				// eslint-disable-next-line no-await-in-loop
-				const status = await this.lessonCopyService.copyLesson({
-					originalLesson,
-					user: params.user,
-					destinationCourse: params.destinationCourse,
-				});
-				elements.push(status);
-				const lessonBardElement = BoardElement.FromLesson(status.copyEntity as Lesson);
-				references.push(lessonBardElement);
-			}
-		}
+		const boardElements: BoardElement[] = originalBoard.getElements();
+		const elements: CopyStatus[] = await this.copyBoardElements(boardElements, user, destinationCourse);
+		const references: BoardElement[] = this.extractReferences(elements);
 
-		const copy = new Board({ references, course: params.destinationCourse });
-		const status = {
+		let boardCopy: Board = new Board({ references, course: destinationCourse });
+		let status: CopyStatus = {
 			title: 'board',
 			type: CopyElementType.BOARD,
 			status: this.copyHelperService.deriveStatusFromElements(elements),
-			copyEntity: copy,
+			copyEntity: boardCopy,
 			originalEntity: params.originalBoard,
 			elements,
 		};
 
+		status = this.updateCopiedEmbeddedTasksOfLessons(status);
+		if (status.copyEntity) {
+			boardCopy = status.copyEntity as Board;
+		}
+		await this.boardRepo.save(boardCopy);
+
 		return status;
 	}
+
+	private async copyBoardElements(
+		boardElements: BoardElement[],
+		user: User,
+		destinationCourse: Course
+	): Promise<CopyStatus[]> {
+		const promises: Promise<[number, CopyStatus]>[] = boardElements.map((element, pos) => {
+			if (element.boardElementType === BoardElementType.Task) {
+				const originalTask = element.target as Task;
```

Comment:
A mapper with validation is required for as cast on any place. 
private castToTask(untypedTask: unknown): Task
// validation check
// cast

Reply:
Solved in BC-2417 using Guard Clauses: https://github.com/hpi-schul-cloud/schulcloud-server/pull/3832/files#diff-6b173d0c37fa49fd9f66c2437ab93c459dbaaa3298958c9979e73c3a910019efR76